### PR TITLE
[Gradle] Use IR to compute used coordinates

### PR DESCRIPTION
**Pull request by martinbonnin**
Opening as draft cause I need to cleanup a few things/review some stuff and it's easier in the GitHub UI. Feel free to take a look but it's not 100% polished yet so it might change. Also doc is still TBD

A few notes:
* I ended up not touching too much the Gradle API cause there is brain muscle around it.
* `apolloMetadata` is removed because it doesn't work in some cases ([Kotlin slack discussion](https://slack-chats.kotlinlang.org/t/9140659/hey-quick-question-do-apollousedcoordinates-support-multiple))
* Instead the dependencies have to be set using service APIs:

```kotlin
# feature
apollo {
  service("service") {
    packageName.set("feature")
    dependsOn(project(":schema")
  }
}

# schema
apollo {
  service("service") {
    packageName.set("schema")
    # naming is up for grabs
    isADependencyOf(project(":feature")
  }
}
```

* All the types are now generated in the schema module compared to earlier where some of them could be generated in multiple different modules. This is because of `data builders` and `__Schema` that need to know about all the types that exist globally.
* There's no "checkApolloDuplicates" any more because of the above. If a backlink is not properly set, it will make a compilation error.

TODO:

* [x] documentation updates
* [x] remove experimental kotlinx serialization okio
* [x] remove non-stable kotlinx serialization

### CODE REVIEW

**Comment by martinbonnin** [comment]
This is now ready for review. Sorry for the huge diff, this was a good spaghetti plate 🍝 !

Fixes https://github.com/apollographql/apollo-kotlin/issues/4350 
Fixes https://github.com/apollographql/apollo-kotlin/issues/4517
See https://github.com/apollographql/apollo-kotlin/issues/4039

**Review by BoD — APPROVED** [review]
👏

About the dsl, are you set on `dependsOn` / `isADependencyOf`? We wondered if it could be `upstreamModule` / `downstreamModule` or maybe something else. On my side I think it's fine like this.

File: libraries/apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/Options.kt [hunk 1]
```diff
@@ -325,6 +302,7 @@ class JavaCodegenOptions(
  * Controls how scalar adapters are used in the generated code.
  */
 @JsonClass(generateAdapter = true, generator = "sealed:type")
+@Serializable
```
[the remark this answers is not in the document]
Comment (martinbonnin) [reply libraries/apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/Options.kt:305]: Good catch, I don't think we still use moshi here, I'll remove it. 
PS: I'm planning to remove Moshi altogether in an upcoming PR, make `apollo-compiler` a step closer to being usable with K/N

**Comment by martinbonnin** [comment]
> About the dsl, are you set on dependsOn / isADependencyOf? We wondered if it could be upstreamModule / downstreamModule or maybe something else. On my side I think it's fine like this.

Not really set but no really strong other ideas so I was planning on keeping it like this unless something else emerges.

**Comment by sampengilly** [comment]
Are both `dependsOn` and `isADependencyOf` required to be provided? I'm hoping for the use case of being able to extract a schema module out as an entirely separate project and publishing it for use across multiple projects. Having that strong two way link seems a bit strange

**Comment by sampengilly** [comment]
From further reading of the changes it looks like it's an either/or situation due to the additional generated classes when depending on the schema module from a feature module correct?

**Pull request merged**